src/riak_kv_entropy_manager.erl

```diff
@@ -507,47 +537,64 @@ add_hashtree_pid(true, Index, Pid, State=#state{trees=Trees, trees_version=VTree
             State4
     end.
 
--spec do_get_lock(any(),pid(),state())
+-spec do_get_lock(any(),pid(), non_neg_integer(), state())
                  -> {ok | max_concurrency | build_limit_reached, state()}.
-do_get_lock(Type, Pid, State=#state{locks=Locks}) ->
+do_get_lock(Type, Pid, Index, State=#state{locks=Locks}) ->
     Concurrency = app_helper:get_env(riak_kv,
                                      anti_entropy_concurrency,
                                      ?DEFAULT_CONCURRENCY),
     case length(Locks) >= Concurrency of
         true ->
             {max_concurrency, State};
         false ->
-            case check_lock_type(Type, State) of
-                {ok, State2} ->
+            case check_lock_type(Type, Locks) of
+                ok ->
                     Ref = monitor(process, Pid),
-                    State3 = State2#state{locks=[{Pid,Ref}|Locks]},
-                    {ok, State3};
+                    NewLock = #lock{pid = Pid, ref = Ref, index = Index, type = Type},
+                    State2 = State#state{locks=[NewLock | Locks]},
+                    {ok, State2};
                 Error ->
                     {Error, State}
             end
     end.
 
+check_lock_type(build, Locks) ->
+    Limit = app_helper:get_env(riak_kv, anti_entropy_build_concurrency, ?DEFAULT_BUILD_CONCURRENCY),
+    NumLocks = get_nr_of_locks(build, Locks) + get_nr_of_locks(upgrade, Locks),
+    case NumLocks >= Limit of
+        true ->
+            build_limit_reached;
+        false ->
+            ok
+    end;
+check_lock_type(_Type, _Locks) ->
+    ok.
+
+get_nr_of_locks(Type, Locks) ->
```

Comment:
This feels like a premature optimization to me. This function shouldn't be called that often (or on a critical path), and I do feel that the simple one-line implementation we have now is much clearer and more readable than a fold would be.

If we had lists:count I would gladly use that, though. Perhaps we should write our own implementation of it and add it to a utility module somewhere. Thoughts?